File: capacity-club-api/src/domain-modules/exercise-training/model/entity/exercise-training.entity.ts
/* eslint-disable @typescript-eslint/no-unused-vars */
import { BaseEntity } from '@common/model';
import { TrainingCircuit } from 'domain-modules/training-circuit/model';
import { PrimaryColumn, Column, ManyToOne, JoinColumn, Entity } from 'typeorm';
import { TrainingIntensity } from '../enum';
import { ExerciseData } from 'domain-modules/exercise-data/model';
import { UniqueId, uniqueIdTransformer } from '@common/model/unique-id';
import { Transform } from 'class-transformer';
import { ApiProperty } from '@nestjs/swagger';

/**
 * Entity representing a specific exercise training session.
 * This entity allows coaches to adapt the general exercise data to a specific training session.
 */
@Entity()
export class ExerciseTraining extends BaseEntity {
  /**
   * Unique identifier for the exercise training.
   * Uses a custom transformer for serialization.
   */
  @PrimaryColumn('varchar')
  @Transform(uniqueIdTransformer.to, { toClassOnly: true })
  @Transform(uniqueIdTransformer.from, { toPlainOnly: true })
  exercise_training_id: UniqueId;
  @Column({ nullable: false })
  nb_reps: number;
  @Column('varchar', { nullable: true })
  intensity: string;
  @Column('varchar', { nullable: false, default: TrainingIntensity.RPE })
  intensityType: TrainingIntensity;

  @ManyToOne((type) => ExerciseData, (data) => data.exercise_training_list, {
    lazy: true,
  })
  @JoinColumn({
    referencedColumnName: 'exercise_data_id',
    name: 'exercise_data_id',
  })
  @ApiProperty({ type: () => ExerciseData })
  exercise_data: Promise<ExerciseData>;

  @ManyToOne(
    (type) => TrainingCircuit,
    (training_circuit) => training_circuit.exercise_training_list,
    {
      lazy: true,
    },
  )
  @JoinColumn({
    referencedColumnName: 'training_circuit_id',
    name: 'training_circuit_id',
  })
  @ApiProperty({ type: () => TrainingCircuit })
  training_circuit: Promise<TrainingCircuit>;
}

File: capacity_club_mobile_app/lib/domain-features/domain/exercise-training/exercise-data/entity/exercise_training_entity.dart
import 'package:capacity_club_mobile_app/core/model/entities/unique_id.dart';
import 'package:capacity_club_mobile_app/domain-features/domain/exercise-data/entity/exercise_data_entity.dart';
import 'package:capacity_club_mobile_app/domain-features/domain/training-circuit/entity/training_circuit_entity.dart';
import 'package:capacity_club_mobile_app/domain-features/data/exercise-training/model/enum/training_intensity.dart';

class ExerciseTrainingEntity {
  final UniqueId exercise_training_id;
  final int nb_reps;
  final String intensity;
  final TrainingIntensity intensityType;
  final ExerciseDataEntity exercise_data;
  final bool is_empty;

  ExerciseTrainingEntity(
      {required this.exercise_training_id,
      required this.nb_reps,
      required this.intensity,
      required this.intensityType,
      required this.exercise_data,
      required this.is_empty});

  factory ExerciseTrainingEntity.empty() {
    return ExerciseTrainingEntity(
        exercise_training_id: UniqueId(''),
        nb_reps: 0,
        intensity: '',
        intensityType: TrainingIntensity.none,
        exercise_data: ExerciseDataEntity.empty(),
        is_empty: true);
  }
}
